OpenRA.Game/Primitives/RotatingBuffer.cs

```diff
@@ -0,0 +1,141 @@
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+
+namespace OpenRA.Primitives
+{
+	/// <summary>Fixed size rorating buffer backed by an array.</summary>
+	public class RotatingBuffer<T> : ICollection<T>, IEnumerable<T>
+	{
+		readonly IComparer<T> comparer;
+		readonly T[] values;
+		int start;
+		int count;
+
+		public int Capacity => values.Length;
+		public int Count => count;
+		public bool IsReadOnly => false;
+
+		public RotatingBuffer(int capacity, IComparer<T> comparer)
+		{
+			this.comparer = comparer;
+			values = new T[capacity];
+			start = 0;
+			count = 0;
+		}
+
+		public RotatingBuffer(int capacity)
+			: this(capacity, Comparer<T>.Default) { }
+
+		public void Add(T value)
+		{
+			values[(start + count) % values.Length] = value;
+			if (count < values.Length)
+				count++;
+			else
+				start = (start + 1) % values.Length;
+		}
+
+		public void Clear()
+		{
+			Array.Clear(values, 0, values.Length);
+			count = 0;
+		}
+
+		public bool Contains(T value)
+		{
+			var capacity = values.Length;
+			var end = start + count;
+			for (var i = start; i < end; ++i)
+				if (comparer.Compare(values[i % capacity], value) == 0)
+					return true;
+
+			return false;
+		}
+
+		public void CopyTo(T[] array, int arrayIndex)
+		{
+			if (array == null)
+				throw new ArgumentNullException("Target array is null");
+
+			if (arrayIndex < 0)
+				throw new ArgumentNullException("Index is less than 0");
+
+			if (arrayIndex + count >= array.Length)
+				throw new ArgumentException("Invalid array capactiy");
+
+			var di = arrayIndex;
+			var end = start + count;
+			var capacity = values.Length;
+			for (var i = start; i < end; ++i)
+				array[di++] = values[i % capacity];
+		}
+
+		public bool Remove(T value)
+		{
+			var capacity = values.Length;
+			var end = start + count;
+			for (var i = start; i < end; ++i)
+				if (comparer.Compare(values[i % capacity], value) == 0)
+				{
+					end--;
+					for (var j = i; j < end; ++j)
+						values[j % capacity] = values[(j + 1) % capacity];
+
+					count--;
+					return true;
+				}
+
+			return false;
+		}
+
+		public T this[int pos]
+		{
+			get => values[(start + pos) % values.Length];
+
+			set
+			{
+				if (pos >= count)
+					throw new ArgumentException($"Index out of bounds: {pos}");
+
+				values[(start + pos) % values.Length] = value;
+			}
+		}
+
+		public T First()
+		{
+			if (count == 0)
+				throw new ArgumentException("Empty buffer");
+
+			return values[start];
+		}
+
+		public T Last()
+		{
+			if (count == 0)
+				throw new ArgumentException("Empty buffer");
+
+			return values[(start + count - 1) % values.Length];
+		}
+
+		public IEnumerator<T> GetEnumerator()
```

Comment:
It looks like this won't behave correctly if the underlying collection is modified between enumerator yields. Other collections check for this and throw an InvalidOperationException("Collection was modified; enumeration operation may not execute").

Reply:
It would still always return the last element. But yes, the user would not be aware.

The down side is the need to check if the start and size changed each access. But if it's a generic container, then it is a good thing to do.